File: Algorithms/198. House Robber/golang/solution.optimized-dp.go
func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	}
	var N int = len(nums)
	var robNext, robNextPlusOne int = nums[N-1], 0
	for i := N - 2; i >= 0; i-- {
		if robNext > robNextPlusOne+nums[i] {
			robNext, robNextPlusOne = robNext, robNext
		} else {
			robNext, robNextPlusOne = robNextPlusOne+nums[i], robNext
		}
	}
	return robNext
}

File: Algorithms/198. House Robber/javascript/solution.optimized-dp.js
/**
 * @param {number[]} nums
 * @return {number}
 */
const rob = function (nums) {
    if (!nums) {
        return 0;
    }

    const N = nums.length;
    let robNextPlusOne = 0,
        robNext = nums[N - 1];

    for (let i = N - 2; i >= 0; i--) {
        const current = Math.max(robNext, robNextPlusOne + nums[i]);
        robNextPlusOne = robNext;
        robNext = current;
    }
    return robNext;
};
